See https://chgibb.github.io/PHATDocs/ for documentation and releases.

![PHAT Tool Bar](https://chgibb.github.io//PHATDocs/docs/releases/0.1.0-beta.1/covHPV16white.png)

The Pathogen-Host Analysis Tool (PHAT) is an application for processing and analyzing next-generation sequencing (NGS) data as it relates to relationships between pathogen and host organisms. PHAT provides quality control (QC) reporting on sequence files, alignment of sequence files against reference files, single-nucleotide polymorphism (SNP) prediction, linear and circular alignment viewing, and Excel and comma separated values (CSV) output.

PHAT is under development in the Zehbe Lab ([http://zehbelab.weebly.com/](http://zehbelab.weebly.com/)) at the Thunder Bay Regional Health Research Institute (TBRHRI) and Lakehead University (LU) under the supervison of Dr. Ingeborg Zehbe. This work is supported by a Natural Sciences and Engineering Research Council of Canada (NSERC) Discovery grant to Dr. Ingeborg Zehbe (#RGPIN-2015-03855) and a NSERC Alexander Graham Bell Canada Graduate Scholarship-Doctoral (CGS-D) to Robert Jackson (#454402-2014).  

**Project Supervisor/Principal Investigator:** Dr. Ingeborg Zehbe, LU/TBRHRI Research Chair & Senior Scientist    
**Project Lead/Biologist:** Robert Jackson, PhD Biotech Candidate    
**Project Lead/Software Developer:** Chris Gibb, HBSc Comp Sci Student  
**Team Members:** Mitchell Pynn, Shane Liu, Jeremy Braun, Nick Catanzaro, Zachary Moorman

**How to Cite PHAT (pre-print):**

Gibb CM, Jackson R, Mohammed S, Fiaidhi J, Zehbe I. Pathogen-Host Analysis Tool (PHAT): an Integrative Platform to Analyze Pathogen-Host Relationships in Next-Generation Sequencing Data. bioRxiv. https://doi.org/10.1101/178327